core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java

```diff
@@ -0,0 +1,99 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.sql.planner;
+
+import com.google.common.collect.ImmutableMap;
+import io.trino.Session;
+import io.trino.metadata.Metadata;
+import io.trino.operator.scalar.ArrayDistinctFunction;
+import io.trino.operator.scalar.ArraySortFunction;
+import io.trino.spi.type.Type;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.ExpressionRewriter;
+import io.trino.sql.tree.ExpressionTreeRewriter;
+import io.trino.sql.tree.FunctionCall;
+import io.trino.sql.tree.NodeRef;
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.SymbolReference;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public final class ArraySortAfterArrayDistinctRewriter
+{
+    public static Expression rewrite(Expression expression, Map<NodeRef<Expression>, Type> expressionTypes, Metadata metadata)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(expressionTypes, metadata), expression);
+    }
+
+    private ArraySortAfterArrayDistinctRewriter() {}
+
+    public static Expression rewrite(Expression expression, Session session, Metadata metadata, TypeAnalyzer typeAnalyzer, TypeProvider typeProvider)
+    {
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+
+        if (expression instanceof SymbolReference) {
+            return expression;
+        }
+        Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, typeProvider, expression);
+
+        return rewrite(expression, expressionTypes, metadata);
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private final Map<NodeRef<Expression>, Type> expressionTypes;
+        private final Metadata metadata;
+
+        public Visitor(Map<NodeRef<Expression>, Type> expressionTypes, Metadata metadata)
+        {
+            this.expressionTypes = ImmutableMap.copyOf(requireNonNull(expressionTypes, "expressionTypes is null"));
+            this.metadata = metadata;
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (metadata.getFunctionMetadata(metadata.decodeFunction(node.getName())).getActualName().equals(ArrayDistinctFunction.NAME)
+                    && node.getArguments().size() > 0
```

Comment:
The `array_distinct` function should have exactly one argument. You could add a checkState here.
At least, change this condition to `node.getArguments().size() == 1` so that you bail out if there are additional arguments, because they're not handled in the following code.